subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/DependencyMetadataType.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.model;
+
+public enum DependencyMetadataType {
+    TRADITIONAL_DEPENDENCY,
```

Comment:
:o: I wonder if it's the right model here. It rather looks like we only have a combination of 2 things:

- is it a constraint or a dependency?
- does it assimilate the contraints or not?

One combination is, _for now_, invalid (constraint + assimilate) but it looks like this should really be a boolean, nothing more.

Reply:
Yes I did this change to avoid the invalid state _constraint + assimilate_. It does not make sense as a constraint alone does not pull in anything. There is nothing to assemble. That's why you also can not define this on constraints in the DSL. Also technically the constraint can become a hard edge in our engine, it does not make sense conceptually. So this combination will never become a valid state.

For me the issue is that we have one implementation for dependencies and for constraints internally here, which should really be more separated. I think the enum makes it a bit more clearer that we have these different types of edges, which could also be more separated in the implementation at some point.

But I don't have a strong opinion. And what you wrote above about refactoring support speaks for the "two booleans" solution.